File: tests/typesafety/typescript/carrier_inheritance.ts
// expect: passes

abstract class Carrier {
  abstract quotePence(kilograms: number): number;
}

class RoyalPost extends Carrier {
  quotePence(kilograms: number): number {
    return Math.floor(kilograms * 120);
  }
}

export function cheapest(carriers: Carrier[], kilograms: number): number {
  return Math.min(...carriers.map((carrier) => carrier.quotePence(kilograms)));
}

export const quote = cheapest([new RoyalPost()], 9.5);

File: tests/typesafety/python/carrier_inheritance.py
# expect: passes
# title: Every carrier must inherit from the base class, even the one you cannot edit
class Carrier:
    def quote_pence(self, kilograms: float) -> int:
        raise NotImplementedError


class RoyalPost(Carrier):
    def quote_pence(self, kilograms: float) -> int:
        return int(kilograms * 120)


def cheapest(carriers: list[Carrier], kilograms: float) -> int:
    return min(carrier.quote_pence(kilograms) for carrier in carriers)
